examples/UNet/massachussets_road_segm.py

```diff
@@ -0,0 +1,340 @@
+    # if you want, use my pretained weights (got around 96% accuracy and a loss of 0.11 using excessive data
+    # augmentation)
+    # https://www.dropbox.com/s/t6juf6o2ix7dntk/UNet_roadSegmentation_Params.zip?dl=0
+    '''with open("UNet_params_ep0.pkl", 'r') as f:
+        params = cPickle.load(f)
+        lasagne.layers.set_all_param_values(output_layer, params)'''
+
+    x_sym = T.tensor4()
+    seg_sym = T.ivector()
+    w_sym = T.vector()
+
+    # add some weight decay
+    l2_loss = lasagne.regularization.regularize_network_params(output_layer, lasagne.regularization.l2) * 1e-4
+
+    # the distinction between prediction_train and test is important only if we enable dropout
+    prediction_train = lasagne.layers.get_output(output_layer, x_sym, deterministic=False)
+    # we could use a binary loss but I stuck with categorical crossentropy so that less code has to be changed if your
+    # application has more than two classes
+    loss = lasagne.objectives.categorical_crossentropy(prediction_train, seg_sym)
+    loss *= w_sym
+    loss = loss.mean()
+    loss += l2_loss
+    acc_train = T.mean(T.eq(T.argmax(prediction_train, axis=1), seg_sym), dtype=theano.config.floatX)
+
+    prediction_test = lasagne.layers.get_output(output_layer, x_sym, deterministic=True)
+    loss_val = lasagne.objectives.categorical_crossentropy(prediction_test, seg_sym)
+
+    # we multiply our loss by a weight map. In this example the weight map only increases the loss for road pixels and
+    # decreases the loss for other pixels. We do this to ensure that the network puts more focus on getting the roads
+    # right
+    loss_val *= w_sym
+    loss_val = loss_val.mean()
+    loss_val += l2_loss
+    acc = T.mean(T.eq(T.argmax(prediction_test, axis=1), seg_sym), dtype=theano.config.floatX)
+
+    # learning rate has to be a shared variablebecause we decrease it with every epoch
+    params = lasagne.layers.get_all_params(output_layer, trainable=True)
+    learning_rate = theano.shared(np.float32(0.001))
+    updates = lasagne.updates.adam(loss, params, learning_rate=learning_rate)
+
+    train_fn = theano.function([x_sym, seg_sym, w_sym], [loss, acc_train], updates=updates)
+    val_fn = theano.function([x_sym, seg_sym, w_sym], [loss_val, acc])
+    # have a look at plot_some_results if you want  to see how to use pred_fn
+    pred_fn = theano.function([x_sym], prediction_test)
+
+    # this is np.sum(target_train == 0) and np.sum(target_train == 1). No need to compute this every time
+    class_frequencies = np.array([2374093357., 118906643.])
+    # we are taking the log here because we want the net to focus more on the road pixels but not too much (otherwise
+    # it would not be penalized enough for missclassifying terrain pixels which results in too many false positives)
+    class_weights = np.log(class_frequencies[[1,0]])
+    class_weights = class_weights / np.sum(class_weights) * 2.
+    class_weights = class_weights.astype(np.float32)
```

Comment:
If you move this above the definition of the loss, you can just do:
```
w_sym = class_weights[seg_sym]
```
Then you don't have to compute and pass the weights to the training and validation function, they'll compute the weights from the targets themselves.